[MultiRep HEIC] Ensure <attachment> creation upon insertion

pxlcoder opened the pull request:
#### 1a21a036fbffe9356104277a52b78df35801cf1c
<pre>
[MultiRep HEIC] Ensure &lt;attachment&gt; creation upon insertion
<a href="https://bugs.webkit.org/show_bug.cgi?id=268243">https://bugs.webkit.org/show_bug.cgi?id=268243</a>
<a href="https://rdar.apple.com/121767827">rdar://121767827</a>

Reviewed by Wenson Hsieh.

Ensure that the HEIC &lt;source&gt; and PNG &lt;img&gt; are associated with &lt;attachment&gt;s
and are registered as `_WKAttachment`s in the UI process.

In order to support this functionality, refactor attachment code to support
associating &lt;source&gt; elements with an &lt;attachment&gt; element, in addition to the
existing support for &lt;img&gt; elements.

`AttachmentAssociatedElement` is introduced to share code between elements the
two elements that can be associated with &lt;attachment&gt;. Additionally rename all
methods that refer to an &quot;enclosing image&quot; to refer to an &quot;associated element&quot;.

Note that this patch does not make it so that *all* &lt;source&gt; elements are
&lt;attachment&gt; associated, only those associated with a multi-representation HEIC.
This is because, in general, not all &lt;source&gt; elements have loaded content.

* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::registerAttachmentIdentifier):
* Source/WebCore/dom/Document.h:
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::asAttachmentAssociatedElement):
* Source/WebCore/dom/Element.h:
* Source/WebCore/editing/Editor.cpp:
(WebCore::Editor::registerAttachmentIdentifier):
(WebCore::Editor::notifyClientOfAttachmentUpdates):
* Source/WebCore/editing/Editor.h:
* Source/WebCore/editing/cocoa/EditorCocoa.mm:
(WebCore::Editor::insertMultiRepresentationHEIC):

Create and register &lt;attachment&gt; elements for the inserted &lt;source&gt; and &lt;img&gt;.

* Source/WebCore/html/AttachmentAssociatedElement.cpp: Added.
(WebCore::AttachmentAssociatedElement::setAttachmentElement):
(WebCore::AttachmentAssociatedElement::attachmentElement const):
(WebCore::AttachmentAssociatedElement::attachmentIdentifier const):
(WebCore::AttachmentAssociatedElement::didUpdateAttachmentIdentifier):
(WebCore::AttachmentAssociatedElement::copyAttachmentAssociatedPropertiesFromElement):
(WebCore::AttachmentAssociatedElement::cloneAttachmentAssociatedElementWithoutAttributesAndChildren):
* Source/WebCore/html/AttachmentAssociatedElement.h: Added.

Introduce `AttachmentAssociatedElement` to share attachment-related code between
&lt;img&gt; and &lt;source&gt; elements.

(WebCore::AttachmentAssociatedElement::ref const):
(WebCore::AttachmentAssociatedElement::deref const):
* Source/WebCore/html/HTMLAttachmentElement.cpp:
(WebCore::HTMLAttachmentElement::getAttachmentIdentifier):
(WebCore::HTMLAttachmentElement::ensureUniqueIdentifier):
(WebCore::HTMLAttachmentElement::setUniqueIdentifier):
(WebCore::HTMLAttachmentElement::associatedElement const):
(WebCore::HTMLAttachmentElement::associatedElementType const):
(WebCore::HTMLAttachmentElement::updateAttributes):
(WebCore::HTMLAttachmentElement::updateAssociatedElementWithData):
(WebCore::HTMLAttachmentElement::enclosingImageElement const): Deleted.
(WebCore::HTMLAttachmentElement::updateEnclosingImageWithData): Deleted.
* Source/WebCore/html/HTMLAttachmentElement.h:
* Source/WebCore/html/HTMLAttachmentElement.idl:

Update `getAttachmentIdentifier` to take an `HTMLElement`, as both
`HTMLSourceElement` and `HTMLImageElement` can be associated with an
`HTMLAttachmentElement`.

* Source/WebCore/html/HTMLImageElement.cpp:
(WebCore::HTMLImageElement::setAttachmentElement):
(WebCore::HTMLImageElement::copyNonAttributePropertiesFromElement):
(WebCore::HTMLImageElement::cloneElementWithoutAttributesAndChildren):
(WebCore::HTMLImageElement::didUpdateAttachmentIdentifier): Deleted.
(WebCore::HTMLImageElement::attachmentElement const): Deleted.
(WebCore::HTMLImageElement::attachmentIdentifier const): Deleted.
* Source/WebCore/html/HTMLImageElement.h:
* Source/WebCore/html/HTMLSourceElement.cpp:
(WebCore::HTMLSourceElement::cloneElementWithoutAttributesAndChildren):
(WebCore::HTMLSourceElement::copyNonAttributePropertiesFromElement):
* Source/WebCore/html/HTMLSourceElement.h:
* Source/WebCore/page/EditorClient.h:
(WebCore::EditorClient::didInsertAttachmentWithIdentifier):
* Source/WebKit/Scripts/webkit/messages.py:
(headers_for_type):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/API/APIAttachment.cpp:
(API::Attachment::associatedElementData const):
(API::Attachment::enclosingImageData const): Deleted.
* Source/WebKit/UIProcess/API/APIAttachment.h:
* Source/WebKit/UIProcess/API/Cocoa/APIAttachmentCocoa.mm:
(API::Attachment::associatedElementData const):
(API::Attachment::associatedElementNSData const):
(API::Attachment::enclosingImageData const): Deleted.
(API::Attachment::enclosingImageNSData const): Deleted.
* Source/WebKit/UIProcess/API/Cocoa/_WKAttachment.h:

Introduce a new `shouldPreserveFidelity` bit on `_WKAttachmentInfo`, to ensure
that clients do not transcode data from a &lt;source&gt; element.

* Source/WebKit/UIProcess/API/Cocoa/_WKAttachment.mm:
(-[_WKAttachmentInfo shouldPreserveFidelity]):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::updateAttachmentAttributes):
(WebKit::WebPageProxy::didInsertAttachmentWithIdentifier):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.messages.in:
* Source/WebKit/UIProcess/mac/WebContextMenuProxyMac.mm:
(WebKit::WebContextMenuProxyMac::setupServicesMenu):
* Source/WebKit/WebProcess/WebCoreSupport/WebEditorClient.cpp:
(WebKit::WebEditorClient::didInsertAttachmentWithIdentifier):
* Source/WebKit/WebProcess/WebCoreSupport/WebEditorClient.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::updateAttachmentAttributes):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:

Canonical link: <a href="https://commits.webkit.org/273731@main">https://commits.webkit.org/273731@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/62ef275a64a17c5dc5af5b5da288dd0682aa5af6

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/36467 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/15406 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/38682 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/39146 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/32732 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/37698 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/17866 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/12496 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/31369 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/37026 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/12990 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/32289 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/11378 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/36594 "Passed tests") | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/11409 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/32531 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/40391 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/33057 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/32872 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/37342 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/11633 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/9496 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/35449 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/13333 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/12075 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/4729 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/12506 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

whsieh commented on Source/WebCore/editing/Editor.cpp:
```diff
@@ -4381,14 +4382,18 @@ void Editor::registerAttachments(Vector<SerializedAttachmentData>&& data)
+        if (!imageElement)
+            return std::nullopt;
+
+        auto* renderer = dynamicDowncast<RenderImage>(imageElement->renderer());
```
Nit - maybe we can turn this into `CheckedPtr` while we're here?

pxlcoder replied:
Done.

whsieh commented on Source/WebCore/editing/cocoa/EditorCocoa.mm:
```diff
@@ -396,25 +396,37 @@ static void maybeCopyNodeAttributesToFragment(const Node& node, DocumentFragment
     ReplaceSelectionCommand::create(document.get(), WTFMove(fragment), ReplaceSelectionCommand::PreventNesting, EditAction::Insert)->apply();
+
+    auto primaryAttachment = HTMLAttachmentElement::create(HTMLNames::attachmentTag, document.get());
+    auto& primaryIdentifier = primaryAttachment->ensureUniqueIdentifier();
```
Nit - I think it's _technically_ unsafe to store these as raw const-refs on the stack, since the code below can trigger layout. Maybe we could change this to just `auto`, or have `ensureUniqueIdentifier` return a `String` instead of a `const String&`.

aprotyas replied:
I think `ensureUniqueIdentifier` should just return a `String`

whsieh replied:
Makes sense! One less footgun.

pxlcoder replied:
Made `ensureUniqueIdentifier` return a `String`.

whsieh commented on Source/WebCore/html/HTMLAttachmentElement.cpp:
```diff
@@ -389,17 +390,21 @@ void HTMLAttachmentElement::invalidateRendering()
+const String& HTMLAttachmentElement::getAttachmentIdentifier(HTMLElement& element)
 {
-    if (auto attachment = image.attachmentElement())
+    auto* attachmentAssociatedElement = element.asAttachmentAssociatedElement();
```
Nit - `RefPtr` and `Ref` below, instead of raw pointers/references.

pxlcoder replied:
Updated to use smart pointers. Note that I also had to update `AttachmentAssociatedElement` to support `ref()` and `defer()`.

whsieh commented on Source/WebCore/html/HTMLSourceElement.cpp:
```diff
@@ -219,4 +219,22 @@ void HTMLSourceElement::addCandidateSubresourceURLs(ListHashSet<URL>& urls) cons
+{
+#if ENABLE(ATTACHMENT_ELEMENT)
+    auto& sourceElement = static_cast<const HTMLSourceElement&>(source);
+    copyAttachmentAssociatedPropertiesFromElement(sourceElement);
```
Could this use `downcast<>` or `dynamicDowncast<>` instead?

pxlcoder replied:
Used `checkedDowncast`.

whsieh commented on Source/WebCore/html/HTMLSourceElement.h:
```diff
@@ -56,6 +62,18 @@ class HTMLSourceElement final : public HTMLElement, public ActiveDOMObject {
+
+    AttachmentAssociatedElement* asAttachmentAssociatedElement() final { return this; }
+
+    AttachmentAssociatedElementType attachmentAssociatedElementType() const final { return AttachmentAssociatedElementType::Source; };
```
Nit - no trailing `;`

pxlcoder replied:
Done.

whsieh commented on Source/WebKit/UIProcess/API/Cocoa/_WKAttachment.h:
```diff
@@ -45,6 +45,7 @@ WK_CLASS_AVAILABLE(macos(10.14), ios(12.0))
 @property (nonatomic, readonly, nullable) NSString *filePath;
 @property (nonatomic, readonly, nullable) NSData *data;
 @property (nonatomic, readonly, nullable) NSFileWrapper *fileWrapper;
+@property (nonatomic, readonly) BOOL preserveFidelity;
```
Nit - maybe `-preservesFidelity` or `-shouldPreserveFidelity`, to make it clear that it's a property? (As opposed to an action method, like `-[_WKAttachment preserveFidelity]`).

I think we might _technically_ need availability declarations too.

pxlcoder replied:
Went with `shouldPreserveFidelity` and added availability declarations.

aprotyas commented on Source/WebCore/page/EditorClient.h:
```diff
@@ -88,7 +92,7 @@ class EditorClient : public CanMakeWeakPtr<EditorClient>, public CanMakeCheckedP
     virtual void registerAttachmentIdentifier(const String& /* identifier */) { }
     virtual void cloneAttachmentData(const String& /* fromIdentifier */, const String& /* toIdentifier */) { }
-    virtual void didInsertAttachmentWithIdentifier(const String& /* identifier */, const String& /* source */, bool /* hasEnclosingImage */) { }
+    virtual void didInsertAttachmentWithIdentifier(const String& /* identifier */, const String& /* source */, AttachmentAssociatedElementType /* associatedElementType */) { }
```
Nit; the comment doesn't add much info, but I see that's the precedence in this file.

aprotyas commented on Source/WebKit/UIProcess/API/Cocoa/_WKAttachment.h:
```diff
@@ -45,6 +45,7 @@ WK_CLASS_AVAILABLE(macos(10.14), ios(12.0))
 @property (nonatomic, readonly, nullable) NSString *filePath;
 @property (nonatomic, readonly, nullable) NSData *data;
 @property (nonatomic, readonly, nullable) NSFileWrapper *fileWrapper;
+@property (nonatomic, readonly) BOOL preserveFidelity;
```
I was expecting availability declarations on this

pxlcoder replied:
Added.

aprotyas commented on Source/WebKit/UIProcess/API/Cocoa/_WKAttachment.mm:
```diff
@@ -109,6 +109,11 @@ - (NSString *)contentType
     return _utiType.get();
 }
 
+- (BOOL)preserveFidelity
```
Nit: `preservesFidelity` might be a better name? It could be read like a verb...

pxlcoder replied:
Went with `shouldPreserveFidelity`.

aprotyas reviewed (commented):
LGTM with some minor comments

pxlcoder committed:
[MultiRep HEIC] Ensure <attachment> creation upon insertion
https://bugs.webkit.org/show_bug.cgi?id=268243
rdar://121767827

Reviewed by Wenson Hsieh.

Ensure that the HEIC <source> and PNG <img> are associated with <attachment>s
and are registered as `_WKAttachment`s in the UI process.

In order to support this functionality, refactor attachment code to support
associating <source> elements with an <attachment> element, in addition to the
existing support for <img> elements.

`AttachmentAssociatedElement` is introduced to share code between elements the
two elements that can be associated with <attachment>. Additionally rename all
methods that refer to an "enclosing image" to refer to an "associated element".

Note that this patch does not make it so that *all* <source> elements are
<attachment> associated, only those associated with a multi-representation HEIC.
This is because, in general, not all <source> elements have loaded content.

* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::registerAttachmentIdentifier):
* Source/WebCore/dom/Document.h:
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::asAttachmentAssociatedElement):
* Source/WebCore/dom/Element.h:
* Source/WebCore/editing/Editor.cpp:
(WebCore::Editor::registerAttachmentIdentifier):
(WebCore::Editor::notifyClientOfAttachmentUpdates):
* Source/WebCore/editing/Editor.h:
* Source/WebCore/editing/cocoa/EditorCocoa.mm:
(WebCore::Editor::insertMultiRepresentationHEIC):

Create and register <attachment> elements for the inserted <source> and <img>.

* Source/WebCore/html/AttachmentAssociatedElement.cpp: Added.
(WebCore::AttachmentAssociatedElement::setAttachmentElement):
(WebCore::AttachmentAssociatedElement::attachmentElement const):
(WebCore::AttachmentAssociatedElement::attachmentIdentifier const):
(WebCore::AttachmentAssociatedElement::didUpdateAttachmentIdentifier):
(WebCore::AttachmentAssociatedElement::copyAttachmentAssociatedPropertiesFromElement):
(WebCore::AttachmentAssociatedElement::cloneAttachmentAssociatedElementWithoutAttributesAndChildren):
* Source/WebCore/html/AttachmentAssociatedElement.h: Added.

Introduce `AttachmentAssociatedElement` to share attachment-related code between
<img> and <source> elements.

(WebCore::AttachmentAssociatedElement::ref const):
(WebCore::AttachmentAssociatedElement::deref const):
* Source/WebCore/html/HTMLAttachmentElement.cpp:
(WebCore::HTMLAttachmentElement::getAttachmentIdentifier):
(WebCore::HTMLAttachmentElement::ensureUniqueIdentifier):
(WebCore::HTMLAttachmentElement::setUniqueIdentifier):
(WebCore::HTMLAttachmentElement::associatedElement const):
(WebCore::HTMLAttachmentElement::associatedElementType const):
(WebCore::HTMLAttachmentElement::updateAttributes):
(WebCore::HTMLAttachmentElement::updateAssociatedElementWithData):
(WebCore::HTMLAttachmentElement::enclosingImageElement const): Deleted.
(WebCore::HTMLAttachmentElement::updateEnclosingImageWithData): Deleted.
* Source/WebCore/html/HTMLAttachmentElement.h:
* Source/WebCore/html/HTMLAttachmentElement.idl:

Update `getAttachmentIdentifier` to take an `HTMLElement`, as both
`HTMLSourceElement` and `HTMLImageElement` can be associated with an
`HTMLAttachmentElement`.

* Source/WebCore/html/HTMLImageElement.cpp:
(WebCore::HTMLImageElement::setAttachmentElement):
(WebCore::HTMLImageElement::copyNonAttributePropertiesFromElement):
(WebCore::HTMLImageElement::cloneElementWithoutAttributesAndChildren):
(WebCore::HTMLImageElement::didUpdateAttachmentIdentifier): Deleted.
(WebCore::HTMLImageElement::attachmentElement const): Deleted.
(WebCore::HTMLImageElement::attachmentIdentifier const): Deleted.
* Source/WebCore/html/HTMLImageElement.h:
* Source/WebCore/html/HTMLSourceElement.cpp:
(WebCore::HTMLSourceElement::cloneElementWithoutAttributesAndChildren):
(WebCore::HTMLSourceElement::copyNonAttributePropertiesFromElement):
* Source/WebCore/html/HTMLSourceElement.h:
* Source/WebCore/page/EditorClient.h:
(WebCore::EditorClient::didInsertAttachmentWithIdentifier):
* Source/WebKit/Scripts/webkit/messages.py:
(headers_for_type):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/API/APIAttachment.cpp:
(API::Attachment::associatedElementData const):
(API::Attachment::enclosingImageData const): Deleted.
* Source/WebKit/UIProcess/API/APIAttachment.h:
* Source/WebKit/UIProcess/API/Cocoa/APIAttachmentCocoa.mm:
(API::Attachment::associatedElementData const):
(API::Attachment::associatedElementNSData const):
(API::Attachment::enclosingImageData const): Deleted.
(API::Attachment::enclosingImageNSData const): Deleted.
* Source/WebKit/UIProcess/API/Cocoa/_WKAttachment.h:

Introduce a new `shouldPreserveFidelity` bit on `_WKAttachmentInfo`, to ensure
that clients do not transcode data from a <source> element.

* Source/WebKit/UIProcess/API/Cocoa/_WKAttachment.mm:
(-[_WKAttachmentInfo shouldPreserveFidelity]):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::updateAttachmentAttributes):
(WebKit::WebPageProxy::didInsertAttachmentWithIdentifier):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.messages.in:
* Source/WebKit/UIProcess/mac/WebContextMenuProxyMac.mm:
(WebKit::WebContextMenuProxyMac::setupServicesMenu):
* Source/WebKit/WebProcess/WebCoreSupport/WebEditorClient.cpp:
(WebKit::WebEditorClient::didInsertAttachmentWithIdentifier):
* Source/WebKit/WebProcess/WebCoreSupport/WebEditorClient.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::updateAttachmentAttributes):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:

Canonical link: https://commits.webkit.org/273731@main
